### PR TITLE
[FSTORE-1238] Add guide for similarity search

### PR DESCRIPTION
**Pull request by kennethmhc**

### CODE REVIEW

File: docs/user_guides/fs/vector_similarity_search.md [hunk 1]
```diff
@@ -0,0 +1,96 @@
+---
```
Comment (robzor92) [inline docs/user_guides/fs/vector_similarity_search.md:1]: Can you add a link from this page https://docs.hopsworks.ai/latest/concepts/mlops/opensearch/ to this guide?

File: docs/user_guides/fs/vector_similarity_search.md [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+description: Users guide about how to use Hopsworks for vector similarity search 
+---
```
Comment (robzor92) [inline docs/user_guides/fs/vector_similarity_search.md:3]: Also please add an entry to this guide on this page: https://docs.hopsworks.ai/3.7/user_guides/fs/

File: docs/user_guides/fs/vector_similarity_search.md [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+First, define the index and embedding features in the vector database. The project index will be used if no index name is provided. 
+
+```aidl
+from hsfs import embedding
+
+# Define the index
+emb = embedding.EmbeddingIndex(index_name=None)
+```
+
+Then, add one or more embedding features to the index. Name and dimension of the embedding features are required for identifying which features should be indexed for k-nearest neighbor (KNN) search. Optionally, you can specify the [similarity function](https://github.com/logicalclocks/feature-store-api/blob/master/python/hsfs/embedding.py#L101).
```
Comment (robzor92) [inline docs/user_guides/fs/vector_similarity_search.md:21]: We shouldn't link directly to the source code. Instead reference the feature-store-api docs directly. 
Look how we do on the API Reference section on this page: https://github.com/logicalclocks/logicalclocks.github.io/blob/7522cbee5395193f7919053307131a7bb30bb527/docs/user_guides/fs/feature_group/create.md

**Review by robzor92 — CHANGES_REQUESTED** [review]

File: docs/user_guides/fs/vector_similarity_search.md [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+2. Choose the features to store in vector store
+
+While it is possible to update feature value in vector store, updating feature value in online store is more efficient. If you have features which are frequently being updated and do not require for filtering, consider storing them separately in a different feature group. As shown in the previous example, `view_cnt` is updated frequently and stored separately. You can then get all the required features by using feature view.
+
+# Next step
+Explore the [notebook example](https://github.com/kennethmhc/news-search-knn-demo/blob/main/news-search-knn-demo.ipynb), demonstrating how to use Hopsworks for implementing a news search application. You can search for news using natural language in the application, powered by the Hopsworks vector database.
```
[the remark this answers is not in the document]
Comment (jimdowling) [reply docs/user_guides/fs/vector_similarity_search.md:96]: Can you add a header for this section (from line 93) -  like
"Performance considerations for Feature Groups with Embeddings"
?

[the remark this answers is not in the document]
Comment (jimdowling) [reply docs/user_guides/fs/vector_similarity_search.md:96]: Also, this line is only correct if news_fg is the label FG:
"You can then get all the required features by using feature view."
So, i would omit it until we get our story straight around embeddings in feature views.

File: docs/user_guides/fs/vector_similarity_search.md (hunk 2 above)
Comment (jimdowling) [reply docs/user_guides/fs/vector_similarity_search.md:3]: Should we not have links to latest?
https://docs.hopsworks.ai/latest/user_guides/fs/

**Review by jimdowling — CHANGES_REQUESTED** [review]
We need to get our story straight around embeddings in feature views.
I don't think the API is clear right now - maybe docs can fix it, but we should discuss.

File: docs/user_guides/fs/vector_similarity_search.md [hunk 5]
```diff
@@ -0,0 +1,96 @@
+---
+description: Users guide about how to use Hopsworks for vector similarity search 
+---
+
+# Introduction
+Vector similarity search is a robust technique enabling the retrieval of similar items based on their embeddings or representations. Its applications range across various domains, from recommendation systems to image similarity and beyond. In Hopsworks, this is facilitated through a vector database, such as Opensearch, which efficiently stores and retrieves relevant embeddings. In this guide, we'll walk you through the process of using Hopsworks for vector similarity search step by step.
+
+# Ingesting Data into the Vector Database
+Hopsworks provides a user-friendly API for writing data to both online and offline feature stores. The example below illustrates the straightforward process of ingesting data into both the vector database and the offline feature store using a single insert method. 
+Currently, Hopsworks supports Opensearch as a vector database.
```
[the remark this answers is not in the document]
Comment (kennethmhc) [reply docs/user_guides/fs/vector_similarity_search.md:10]: It is not configurable right now. I can include the default setting.
```
Method: hnsw
Engine: nmslib
```

File: docs/user_guides/fs/vector_similarity_search.md [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+# Ingesting Data into the Vector Database
+Hopsworks provides a user-friendly API for writing data to both online and offline feature stores. The example below illustrates the straightforward process of ingesting data into both the vector database and the offline feature store using a single insert method. 
+Currently, Hopsworks supports Opensearch as a vector database. 
+
+First, define the index and embedding features in the vector database. The project index will be used if no index name is provided. 
+
+```aidl
+from hsfs import embedding
+
+# Define the index
```
[the remark this answers is not in the document]
Comment (kennethmhc) [reply docs/user_guides/fs/vector_similarity_search.md:17]: I just wanted to show that it is possible to define the index name, and it is not required.

File: docs/user_guides/fs/vector_similarity_search.md [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+```aidl
+# Search neighbor embedding with k=3
+news_fg.find_neighbors(model.encode(news_description), k=3)
+
+# Filter and search
+news_fg.find_neighbors(model.encode(news_description), k=3, filter=news_fg.newstype == "sports")
+```
+
+To retrieve features at a specific time in the past from the offline database for analysis, you can utilize the offline read API to perform time travel.
```
[the remark this answers is not in the document]
Comment (kennethmhc) [reply docs/user_guides/fs/vector_similarity_search.md:54]: The purpose is to show how to read embedding from **offline** feature store

File: docs/user_guides/fs/vector_similarity_search.md [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+
+```aidl
+# Time travel and read from the offline feature store
+news_fg.as_of(time_in_past).read()
+```
+
+## Second Phase Reranking
```
[the remark this answers is not in the document]
Comment (kennethmhc) [reply docs/user_guides/fs/vector_similarity_search.md:61]: Yea the header may be misleading here.

File: docs/user_guides/fs/vector_similarity_search.md [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+
+In some ML applications, second phase reranking of the top k items fetched by first phase filtering is common where extra features are required from other sources after fetching the k nearest items. In practice, it means that an extra step is needed to fetch the features from other feature groups in the online feature store. Hopsworks provides yet another simple read API for this purpose. Users can create a feature view by joining multiple feature groups and fetch all the required features by calling fv.find_neighbors. In the example below, view_cnt from another feature group is also returned to the result.
```
[the remark this answers is not in the document]
Comment (kennethmhc) [reply docs/user_guides/fs/vector_similarity_search.md:63]: The purpose is to say: "You can also use find_neighbors in feature view". Second phase reranking is just a motivate example. It is just one sentence.

File: docs/user_guides/fs/vector_similarity_search.md [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+    topic_name=f"view_fg_{version}_onlinefs"
+)
+
+fv = fs.get_or_create_feature_view(
+    "news_cnt", version=version,
+    query=news_fg.select(["date", "heading", "newstype"]).join(view_fg.select(["view_cnt"])))
+
+fv.find_neighbors(model.encode(news_description), k=5)
+```
+
```
[the remark this answers is not in the document]
Comment (kennethmhc) [reply docs/user_guides/fs/vector_similarity_search.md:80]: > I assume you can only return the features from the vector embedding feature group - not from the other feature groups that make up the feature view.

No fv.find_neighbors return features from other feature group also, `view_fg` is stored in RonDb.
The purpose of fv.find_neighbor is
1. select features from embedding feature group, some features may be just stored for filtering purpose and is not part of the model features.
2. save users' effort by combining 2 steps: 1. getting similar items in feature group and then 2. get feature vector from feature view, into a single method.

File: docs/user_guides/fs/vector_similarity_search.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (kennethmhc) [reply docs/user_guides/fs/vector_similarity_search.md:96]: fv do returns all feature as long as it has the primary key of other feature groups

**Review by robzor92 — APPROVED** [review]

File: docs/user_guides/fs/vector_similarity_search.md [hunk 11]
```diff
@@ -0,0 +1,111 @@
+---
+description: User guide for how to use vector similarity search in Hopsworks
+---
+
+# Introduction
+Vector similarity search is a technique enabling the retrieval of similar items based on their vector embeddings or representations. Its applications range across various domains, from recommendation systems to image similarity and beyond. In Hopsworks, vector similarity search is enabled by extending an online feature group with approximate nearest neighbor search capabilities through a vector database, such as Opensearch. This guide provides a detailed walkthrough on how to leverage Hopsworks for vector similarity search.
```
Comment (jimdowling) [inline docs/user_guides/fs/vector_similarity_search.md:6]: Are we consistent in calling it "vector similarity search" in all the docs?

Comment (jimdowling) [reply docs/user_guides/fs/vector_similarity_search.md:6]: Then i would add -
Vector similarity search (also called similarity search)

Comment (jimdowling) [reply docs/user_guides/fs/vector_similarity_search.md:6]: Or just change it to "vector similarity search" everywhere in the docs.

File: docs/user_guides/fs/vector_similarity_search.md [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+In the following example, we explicitly define an index for the feature group:
+
+```aidl
+from hsfs import embedding
+
+# Specify optionally the index in the vector database
+emb = embedding.EmbeddingIndex(index_name="news_fg")
+```
+
+Then, add one or more embedding features to the index. Name and dimension of the embedding features are required for identifying which features should be indexed for k-nearest neighbor (KNN) search. In this example, we get the dimension of the embedding by taking the length of the value of the `embedding_heading` column in the first row of the dataframe `df`. Optionally, you can specify the similarity function.
```
Comment (jimdowling) [inline docs/user_guides/fs/vector_similarity_search.md:20]: This is not clear:
"Optionally, you can specify the similarity function."
You should add a couple of examples.
Optionally, you can specify the similarity function, for example, cosine or ...

File: docs/user_guides/fs/vector_similarity_search.md [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+```aidl
+# Add embedding feature to the index
+emb.add_embedding("embedding_heading", len(df["embedding_heading"][0]))
+```
+
+Next, you create a feature group with the `embedding_index` and ingest data to the feature group. When the `embedding_index` is provided, the vector database is used as online feature store. That is, all the features in the feature group are stored **exclusively** in the vector database. The advantage of storing all features in the vector database is that it enables similarity search, and filtering for all feature values.
```
Comment (jimdowling) [inline docs/user_guides/fs/vector_similarity_search.md:26]: filtering => push-down filtering

**Review by jimdowling — APPROVED** [review]
Just a few small changes

**Pull request merged**